### PR TITLE
Closes Issue #4 (Progress Bar Feature Implementation)

### PR DESCRIPTION
**Pull request by mehtajinesh**
# Progress Bar Feature Implementation

## Overview
This PR implements progress bar functionality across all slow evaluation functions in the UQLM library, providing users with visual feedback during long-running operations. The implementation uses the `rich` library for beautiful, feature-rich progress bars.

### New Dependencies
- **`rich` (^13.8.0)**: Added as a main dependency for progress bar functionality
- **`tf-keras` (2.19.0)**: Added as a test dependency for testing purposes

#### 1. Response Generator (`uqlm/utils/response_generator.py`)
- **Replaced `tqdm` with `rich.progress`**: Migrated from tqdm to rich for better visual experience
- **Added progress bar support**: New `progress_bar` parameter in `generate_responses()` method
- **Refactored batch processing**: Extracted `_process_batch()` method for cleaner code organization
- **Enhanced progress display**: Shows count per prompt in progress bar description

#### 2. Black Box Scorers
All black box scorers now support progress bars:
- **`uqlm/black_box/bert.py`**: Added progress bar to `evaluate()` method
- **`uqlm/black_box/bleurt.py`**: Added progress bar to `evaluate()` method  
- **`uqlm/black_box/cosine.py`**: Added progress bar to `evaluate()` method
- **`uqlm/black_box/match.py`**: Added progress bar to `evaluate()` method
- **`uqlm/black_box/nli.py`**: Added progress bar to `evaluate()` method

#### 3. Uncertainty Quantifiers
All uncertainty quantifiers now support progress bars:
- **`uqlm/scorers/ensemble.py`**: Added progress bar to `generate_and_score()`, `score()`, and `tune()` methods
- **`uqlm/scorers/entropy.py`**: Added progress bar to `generate_and_score()` method
- **`uqlm/scorers/panel.py`**: Added progress bar to `generate_and_score()` method
- **`uqlm/scorers/white_box.py`**: Added progress bar to `generate_and_score()` method
- **`uqlm/scorers/baseclass/uncertainty.py`**: Updated base class to support progress bar parameter

#### 4. Comprehensive Test Suite (`tests/test_progress_bar.py`)
- **269 lines of new tests**: Complete test coverage for progress bar functionality
- **Async and sync testing**: Tests for both async and synchronous operations
- **Mock implementations**: Proper mocking for LLM interactions
- **Edge case handling**: Tests for empty data, invalid parameters, etc.
- **All scorer types**: Tests for all black box scorers and uncertainty quantifiers

#### New Parameters
All relevant methods now accept an optional `progress_bar` parameter:
```python
# Example usage
scorer.evaluate(responses, sampled_responses, progress_bar=True)
await uqe.generate_and_score(prompts, progress_bar=True)
```

#### Backward Compatibility
- All `progress_bar` parameters default to `False`
- Existing code continues to work without modification
- No breaking changes to existing APIs

### Test Coverage
- **Unit tests**: 269 lines of comprehensive test coverage
- **Integration tests**: Tests with actual scorer implementations
- **Async tests**: Proper async/await testing with pytest-asyncio
- **Mock tests**: Isolated testing with mocked dependencies

### Test Categories
1. **Rich library functionality**: Basic progress bar operations
2. **Response generator**: Async response generation with progress bars
3. **Black box scorers**: All scorer types (Match, Cosine, BERT, etc.)
4. **Uncertainty quantifiers**: Ensemble, entropy, panel, white box
5. **Parameter validation**: Edge cases and invalid inputs
6. **Empty data handling**: Tests with empty datasets

### Files Changed
- **15 files modified**: Core library files and configuration
- **755 insertions, 150 deletions**: Net addition of 605 lines
- **New test file**: Complete test suite for progress bar functionality

### Performance
- **Minimal overhead**: Progress bars only add negligible performance cost when disabled
- **Optional feature**: Users can disable progress bars for maximum performance
- **Rich library**: Efficient progress bar implementation

### Basic Usage
```python
from uqlm.black_box import MatchScorer

scorer = MatchScorer()
# With progress bar
results = scorer.evaluate(responses, sampled_responses, progress_bar=True)
# Without progress bar (default)
results = scorer.evaluate(responses, sampled_responses)
```

### Async Operations
```python
from uqlm.scorers import BlackBoxUQ

uqe = BlackBoxUQ(llm=llm, scorers=["exact_match"])
# With progress bar
results = await uqe.generate_and_score(prompts, progress_bar=True)
```

### Ensemble Tuning
```python
from uqlm.scorers import UQEnsemble

ensemble = UQEnsemble(llm=llm, scorers=["exact_match", "semantic_entropy"])
# With progress bar during tuning
results = await ensemble.tune(prompts, ground_truth, progress_bar=True)
```

### Implementation Approach
1. **Consistent API**: All progress bar parameters follow the same pattern
2. **Rich library**: Modern, feature-rich progress bar implementation
3. **Async support**: Proper async/await integration
4. **Error handling**: Graceful handling of progress bar failures
5. **Performance**: Minimal overhead when disabled

### Code Quality
- **Type hints**: Full type annotation support
- **Documentation**: Updated docstrings with progress bar parameters
- **Error handling**: Robust error handling for edge cases
- **Testing**: Comprehensive test coverage

### For Existing Users
No migration required! All existing code continues to work as-is. Progress bars are opt-in features.

### For New Users
Simply add `progress_bar=True` to any evaluation method for visual feedback:

```python
# Before (still works)
results = scorer.evaluate(responses, sampled_responses)

# After (with progress bar)
results = scorer.evaluate(responses, sampled_responses, progress_bar=True)
```

## 🎨 Visual Improvements

The `rich` library provides:
- **Beautiful progress bars**: Modern, colorful progress indicators
- **Detailed information**: Shows current operation and progress percentage
- **Consistent styling**: Professional appearance across all operations
- **Cross-platform**: Works on all operating systems and terminals

## 🔍 Code Review Checklist

- [x] **Backward compatibility**: No breaking changes
- [x] **Type hints**: All new parameters properly typed
- [x] **Documentation**: Updated docstrings and examples
- [x] **Testing**: Comprehensive test coverage
- [x] **Performance**: Minimal overhead when disabled
- [x] **Error handling**: Robust error handling
- [x] **Code style**: Follows project conventions
- [x] **Dependencies**: Minimal new dependencies

## 🚀 Future Enhancements

Potential future improvements:
- **Custom progress bar styles**: User-configurable progress bar appearance
- **Progress callbacks**: Custom progress update callbacks
- **Nested progress bars**: Support for nested operations
- **Progress persistence**: Save/restore progress for long operations

---

**Branch**: `mihir/progress_bar`  
**Base**: `upstream/mihir/progress_bar`  
**Commits**: 4 commits  
**Files Changed**: 15 files  
**Lines Added**: 755 insertions, 150 deletions

### CODE REVIEW

**Comment by mehtajinesh** [comment]
Snapshot of tests failing without "tf-keras" installation for local testing


image

**Review by dylanbouchard — CHANGES_REQUESTED** [review]
Thanks Jinesh! I love the way this is coming along. It looks great. I would request the following changes:

1. There is an apparent bug that prevents the progress bars from properly completing (screenshot below). Let's resolve this. Let me know if I can help.
2. Lets set default=True for progress bars.
3. Lets remove the existing print statements as they are now redundant.

image

**Comment by mehtajinesh** [comment]
@dylanbouchard  Completed all changes. Please review.

File: uqlm/utils/response_generator.py [hunk 1]
```diff
@@ -123,40 +122,51 @@ def _update_count(self, count: int) -> None:
...
+
+        if progress_bar:
+            with Progress(SpinnerColumn(), TextColumn("[progress.description]{task.description}"), BarColumn(), TextColumn("[progress.percentage]{task.completed}/{task.total}"), TimeElapsedColumn()) as progress:
+                if self.count == 1:
+                    task = progress.add_task("[green]Generating responses...", total=len(prompts_partition))
+                else:
+                    task = progress.add_task(f"[green]Generating candidate responses ({self.count} per prompt)...", total=len(prompts_partition))
+                for prompt_batch in prompts_partition:
+                    await self._process_batch(prompt_batch, duplicated_prompts, generations, batch_size)
+                    progress.update(task, advance=1)
```
Comment (dylanbouchard) [inline uqlm/utils/response_generator.py:141]: this should be `advance=len(prompt_batch) * self.count`

File: uqlm/utils/response_generator.py [hunk 2]
```diff
@@ -123,40 +122,51 @@ def _update_count(self, count: int) -> None:
...
-            # pause if needed
-            if (stop - start < 60) and (batch_size < len(prompts)):
-                time.sleep(61 - stop + start)
+
+        if progress_bar:
+            with Progress(SpinnerColumn(), TextColumn("[progress.description]{task.description}"), BarColumn(), TextColumn("[progress.percentage]{task.completed}/{task.total}"), TimeElapsedColumn()) as progress:
+                if self.count == 1:
+                    task = progress.add_task("[green]Generating responses...", total=len(prompts_partition))
+                else:
+                    task = progress.add_task(f"[green]Generating candidate responses ({self.count} per prompt)...", total=len(prompts_partition))
```
Comment (dylanbouchard) [inline uqlm/utils/response_generator.py:138]: this should be `len(prompts) * self.count`

File: uqlm/utils/response_generator.py [hunk 3]
```diff
@@ -123,40 +122,51 @@ def _update_count(self, count: int) -> None:
...
-            stop = time.time()
-
-            # pause if needed
-            if (stop - start < 60) and (batch_size < len(prompts)):
-                time.sleep(61 - stop + start)
+
+        if progress_bar:
+            with Progress(SpinnerColumn(), TextColumn("[progress.description]{task.description}"), BarColumn(), TextColumn("[progress.percentage]{task.completed}/{task.total}"), TimeElapsedColumn()) as progress:
+                if self.count == 1:
+                    task = progress.add_task("[green]Generating responses...", total=len(prompts_partition))
```
Comment (dylanbouchard) [inline uqlm/utils/response_generator.py:136]: this should be `len(prompts)`

File: uqlm/scorers/ensemble.py [hunk 4]
```diff
@@ -116,7 +116,7 @@ def __init__(
         self._validate_components(scorers)
         self._validate_weights()
 
-    async def generate_and_score(self, prompts: List[str], num_responses: int = 5):
+    async def generate_and_score(self, prompts: List[str], num_responses: int = 5, progress_bar: Optional[bool] = False):
```
Comment (dylanbouchard) [inline uqlm/scorers/ensemble.py:119]: can we make default progress_bar =True?

Comment (dylanbouchard) [reply uqlm/scorers/ensemble.py:119]: analogously for all other user-facing class methods

**Review by dylanbouchard — CHANGES_REQUESTED** [review]
Looks amazing, thanks Jinesh!!

Two changes requested:
1. Progress bars are not properly working for generation steps, minor fixes needed in `ResponseGenerator` class (left comments to show where). 
2. Please update user-facing class methods to have default setting of progress_bar=True

File: uqlm/utils/response_generator.py (hunk 3 above)
Comment (mehtajinesh) [reply uqlm/utils/response_generator.py:136]: Got it. So this will work for each batch rather than for each prompt inside a batch.

File: uqlm/utils/response_generator.py (hunk 1 above)
Comment (mehtajinesh) [reply uqlm/utils/response_generator.py:141]: Done.

Comment (dylanbouchard) [reply uqlm/utils/response_generator.py:141]: actually we don't need this here anymore (as reflected in updated code that I shared)
